pkg/storage/segments/repairer.go

```diff
@@ -0,0 +1,157 @@
+	return &segmentRepairer{oc: oc, ec: ec, pdb: pdb, nodeStats: nodeStats}
+}
+
+// Repair retrieves an at-risk segment and repairs and stores lost pieces on new nodes
+func (s *segmentRepairer) Repair(ctx context.Context, path storj.Path, lostPieces []int32) (err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	// Read the segment's pointer's info from the PointerDB
+	pr, originalNodes, pba, err := s.pdb.Get(ctx, path)
+	if err != nil {
+		return Error.Wrap(err)
+	}
+
+	if pr.GetType() != pb.Pointer_REMOTE {
+		return Error.New("cannot repair inline segment %s", psclient.PieceID(pr.GetInlineSegment()))
+	}
+
+	seg := pr.GetRemote()
+	pid := psclient.PieceID(seg.GetPieceId())
+
+	originalNodes, err = lookupAndAlignNodes(ctx, s.oc, originalNodes, seg)
+	if err != nil {
+		return Error.Wrap(err)
+	}
+
+	// Get the nodes list that needs to be excluded
+	var excludeNodeIDs storj.NodeIDList
+
+	// Count the number of nil nodes thats needs to be repaired
+	totalNilNodes := 0
+
+	healthyNodes := make([]*pb.Node, len(originalNodes))
+
+	// Populate healthyNodes with all nodes from originalNodes except those correlating to indices in lostPieces
+	for i, v := range originalNodes {
+		if v == nil {
+			totalNilNodes++
+			continue
+		}
+
+		excludeNodeIDs = append(excludeNodeIDs, v.Id)
+
+		// If node index exists in lostPieces, skip adding it to healthyNodes
+		if contains(lostPieces, i) {
+			totalNilNodes++
+		} else {
+			healthyNodes[i] = v
+		}
+	}
+
+	// Request Overlay for n-h new storage nodes
+	op := overlay.Options{Amount: totalNilNodes, Space: 0, Excluded: excludeNodeIDs}
+	newNodes, err := s.oc.Choose(ctx, op)
+	if err != nil {
+		return err
+	}
+
+	if totalNilNodes != len(newNodes) {
+		return Error.New("Number of new nodes from overlay (%d) does not equal total nil nodes (%d)", len(newNodes), totalNilNodes)
+	}
+
+	totalRepairCount := len(newNodes)
+
+	// Make a repair nodes list just with new unique ids
+	repairNodes := make([]*pb.Node, len(healthyNodes))
+	for i, vr := range healthyNodes {
+		// Check that totalRepairCount is non-negative
+		if totalRepairCount < 0 {
+			return Error.New("Total repair count (%d) less than zero", totalRepairCount)
+		}
+
+		// Find the nil nodes in the healthyNodes list
+		if vr == nil {
+			// Assign the item in repairNodes list with an item from the newNode list
+			totalRepairCount--
+			repairNodes[i] = newNodes[totalRepairCount]
+		}
+	}
+
+	// Check that all nil nodes have a replacement prepared
+	if totalRepairCount != 0 {
+		return Error.New("Failed to replace all nil nodes (%d). (%d) new nodes not inserted", len(newNodes), totalRepairCount)
+	}
+
+	rs, err := makeRedundancyStrategy(pr.GetRemote().GetRedundancy())
+	if err != nil {
+		return Error.Wrap(err)
+	}
+
+	signedMessage := s.pdb.SignedMessage()
+
+	// Download the segment using just the healthyNodes
+	rr, err := s.ec.Get(ctx, healthyNodes, rs, pid, pr.GetSegmentSize(), pba, signedMessage)
```

Comment:
Should there be a check whether the `healthyNodes` are sufficient to download in the first place?

Reply:
There is such check just inside the `ec.Get` method.

Reply:
Hmm, yeah, was thinking whether extra logging about it might be useful, but yeah, should be sufficient.